learning_observer/learning_observer/google.py

```diff
@@ -404,73 +94,168 @@ def get_error_details(error):
+        if align:
+            lengths = [f['endIndex'] - f['startIndex'] for f in flat]
+            text_chunks = [_force_text_length(chunk, length) for chunk, length in zip(text_chunks, lengths)]
+        text = ''.join(text_chunks)
+
+        if EXTRACT_DEBUG_CHECKS:
+            print("Text length versus Google length:")
+            print(len(text), length)
+            print("We expect these to be off by one, since Google seems to starts at 1 (and Python at 0)")
+            if align:
+                print
+                print("Offsets (these should match):")
+                print(list(zip(itertools.accumulate(map(len, text_chunks)), itertools.accumulate(lengths))))
+
+        return {'text': text}
+
+    @register_cleaner("coursework_submissions", "assigned_docs", GOOGLE_ENDPOINTS)
+    def clean_assignment_docs(google_json):
+        '''
+        Retrieve set of documents per student associated with an assignment
+        '''
+        student_submissions = google_json.get('studentSubmissions', [])
+        for student_json in student_submissions:
+            google_id = student_json[constants.USER_ID]
+            local_id = learning_observer.auth.google_id_to_user_id(google_id)
+            student_json[constants.USER_ID] = local_id
+            docs = [d['driveFile'] for d in learning_observer.util.get_nested_dict_value(student_json, 'assignmentSubmission.attachments', []) if 'driveFile' in d]
+            student_json['documents'] = docs
+            # TODO we should probably remove some of the keys provided
+        return student_submissions
+
+    @learning_observer.prestartup.register_startup_check
+    def connect_to_google_cache():
+        '''Setup cache for requests to the Google API.
+        The cache is currently only used with the `use_google_ajax`
+        feature flag.
+        '''
+        if 'google_routes' not in settings.settings['feature_flags']:
+            return
+
+        for key in ['save_google_ajax', 'use_google_ajax', 'save_clean_ajax', 'use_clean_ajax']:
+            if key in settings.settings['feature_flags']:
+                global cache
+                try:
+                    cache = learning_observer.kvs.KVS.google_cache()
+                except AttributeError:
+                    error_text = 'The google_cache KVS is not configured.\n'\
+                        'Please add a `google_cache` kvs item to the `kvs` '\
+                        'key in `creds.yaml`.\n'\
+                        '```\ngoogle_cache:\n  type: filesystem\n  path: ./learning_observer/static_data/google\n'\
+                        '  subdirs: true\n```\nOR\n'\
+                        '```\ngoogle_cache:\n  type: redis_ephemeral\n  expiry: 600\n```'
+                    raise learning_observer.prestartup.StartupCheck("Google KVS: " + error_text) 
+        
+google_lms = GoogleLMS()
+
+def initialize_google_routes(app):
+    google_lms.initialize_routes(app)
```

Comment:
Not the biggest fan of this code. If we don't need/want an LMS setup, we should forgo setting it up. There ought to be a way to determine which LMSs are present (found via creds.yaml) and only update those. 

This loosely ties into my comment about the KVS_MAP.

Reply:
Agreed.

Reply:
I implemented this but made `Google` routes available regardless because if `Canvas` is selected as the default LMS, we still need the `extract_text_from_google_doc_json` function except I take this function out and put it separately.

However, if other LMSes are added, it will automatically initialize the routes for it and not every single LMS present

Reply:
If each item is going to be of an LMS class, then why do we need this extra function for initialization?

Reply:
Oh, this `extract_text_from_google_doc_json` function is more of a Google Docs API rather than Google Classroom as an LMS